File: dog-3d-app/tools/generate_dog_glb.mjs
import { Document, NodeIO, Accessor } from '@gltf-transform/core';
import * as THREE from 'three';
import path from 'node:path';
import fs from 'node:fs/promises';
import { fileURLToPath } from 'node:url';

// 低ポリの犬GLBをNode.jsで生成するスクリプト
// Three.jsのプリミティブからジオメトリを組み立て、gltf-transformでGLBを書き出す。

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const OUTPUT_PATH = path.join(__dirname, '..', 'public', 'models', 'dog.glb');

const doc = new Document();
const buffer = doc.createBuffer('DogBuffer');
const scene = doc.createScene('DogScene');

function hexToFactor(hex) {
  const color = new THREE.Color(hex);
  return [color.r, color.g, color.b, 1];
}

function createMaterial(name, hex) {
  return doc
    .createMaterial(name)
    .setBaseColorFactor(hexToFactor(hex))
    .setMetallicFactor(0)
    .setRoughnessFactor(0.6);
}

const materials = {
  body: createMaterial('Dog_Body_Mat', 0x8a6547),
  ear: createMaterial('Dog_Ear_Mat', 0x5a3d27),
  nose: createMaterial('Dog_Nose_Mat', 0x1a1a1a),
  eye: createMaterial('Dog_Eye_Mat', 0x101010),
  paw: createMaterial('Dog_Paw_Mat', 0xd8b99b),
};

function accessorFromArray(name, array, type) {
  return doc.createAccessor(name).setType(type).setArray(array).setBuffer(buffer);
}

function geometryToPrimitive(name, geometry, material) {
  const primitive = doc.createPrimitive(name).setMaterial(material);
  const position = geometry.attributes.position;
  if (!position) {
    throw new Error(`${name} に position 属性がありません`);
  }
  primitive.setAttribute(
    'POSITION',
    accessorFromArray(`${name}_POSITION`, new Float32Array(position.array), Accessor.Type.VEC3)
  );

  if (geometry.attributes.normal) {
    primitive.setAttribute(
      'NORMAL',
      accessorFromArray(`${name}_NORMAL`, new Float32Array(geometry.attributes.normal.array), Accessor.Type.VEC3)
    );
  }

  if (geometry.attributes.uv) {
    primitive.setAttribute(
      'TEXCOORD_0',
      accessorFromArray(`${name}_UV`, new Float32Array(geometry.attributes.uv.array), Accessor.Type.VEC2)
    );
  }

  if (geometry.index) {
    const indexArray = geometry.index.array;
    const needsUint32 = geometry.attributes.position.count > 65535;
    const indices = needsUint32 ? new Uint32Array(indexArray) : new Uint16Array(indexArray);
    primitive.setIndices(accessorFromArray(`${name}_INDICES`, indices, Accessor.Type.SCALAR));
  }

  return primitive;
}

function createMesh(name, parts) {
  const mesh = doc.createMesh(name);
  parts.forEach((part, index) => {
    mesh.addPrimitive(geometryToPrimitive(`${name}_part${index}`, part.geometry, part.material));
  });
  return mesh;
}

function setNodeRotationFromEuler(node, x, y, z) {
  const quat = new THREE.Quaternion().setFromEuler(new THREE.Euler(x, y, z));
  node.setRotation([quat.x, quat.y, quat.z, quat.w]);
}

// --- ジオメトリ定義 ---
const bodyGeometry = new THREE.CapsuleGeometry(0.4, 0.9, 6, 12);
bodyGeometry.rotateX(Math.PI / 2);

const headGeometry = new THREE.SphereGeometry(0.35, 16, 12);
const snoutBox = new THREE.BoxGeometry(0.34, 0.22, 0.4);
snoutBox.translate(0, 0.02, 0.38);
const noseSphere = new THREE.SphereGeometry(0.1, 10, 8);
noseSphere.translate(0, 0.03, 0.65);

const eyeSphere = new THREE.SphereGeometry(0.06, 8, 6);
eyeSphere.translate(0, 0.02, 0.46);

const earCone = new THREE.ConeGeometry(0.14, 0.5, 4, 1);
earCone.translate(0, -0.25, 0);

const tailGeometry = new THREE.CylinderGeometry(0.08, 0.11, 0.7, 6, 1);
tailGeometry.rotateX(Math.PI / 2);
tailGeometry.translate(0, 0, -0.35);

const legRadius = 0.1;
const legLength = 0.16;
const legGeometry = new THREE.CapsuleGeometry(legRadius, legLength, 4, 8);
const legHalfHeight = legLength / 2 + legRadius;
legGeometry.translate(0, -legHalfHeight, 0);
const pawHeight = 0.12;
const pawGeometry = new THREE.CylinderGeometry(0.13, 0.16, pawHeight, 6, 1);
pawGeometry.translate(0, -legHalfHeight - pawHeight / 2, 0);

// --- メッシュ作成 ---
const bodyMesh = createMesh('BodyMesh', [{ geometry: bodyGeometry, material: materials.body }]);
const headMesh = createMesh('HeadMesh', [{ geometry: headGeometry, material: materials.body }]);
const snoutMesh = createMesh('SnoutMesh', [
  { geometry: snoutBox, material: materials.body },
  { geometry: noseSphere, material: materials.nose },
]);
const eyeMesh = createMesh('EyeMesh', [{ geometry: eyeSphere, material: materials.eye }]);
const earMesh = createMesh('EarMesh', [{ geometry: earCone, material: materials.ear }]);
const tailMesh = createMesh('TailMesh', [{ geometry: tailGeometry, material: materials.body }]);
const legMesh = createMesh('LegMesh', [
  { geometry: legGeometry, material: materials.body },
  { geometry: pawGeometry, material: materials.paw },
]);

// --- ノード階層 ---
const rootNode = doc.createNode('DogRoot');
scene.addChild(rootNode);

doc.getRoot().setDefaultScene(scene);

const bodyNode = doc.createNode('Body').setMesh(bodyMesh).setTranslation([0, 0.58, 0]);
rootNode.addChild(bodyNode);

const headPivot = doc.createNode('HeadPivot').setTranslation([0, 0.95, 0.68]);
rootNode.addChild(headPivot);
const headNode = doc.createNode('Head').setMesh(headMesh).setTranslation([0, 0.08, 0.24]);
setNodeRotationFromEuler(headNode, -0.05, 0, 0);
const snoutNode = doc.createNode('Snout').setMesh(snoutMesh);
const eyeL = doc.createNode('Eye_L').setMesh(eyeMesh).setTranslation([-0.16, 0.05, 0]);
const eyeR = doc.createNode('Eye_R').setMesh(eyeMesh).setTranslation([0.16, 0.05, 0]);
const earL = doc.createNode('Ear_L').setMesh(earMesh).setTranslation([-0.22, 0.15, -0.12]);
const earR = doc.createNode('Ear_R').setMesh(earMesh).setTranslation([0.22, 0.15, -0.12]);
setNodeRotationFromEuler(earL, -0.4, 0, 0.35);
setNodeRotationFromEuler(earR, -0.4, 0, -0.35);
headPivot.addChild(headNode);
headPivot.addChild(snoutNode);
headPivot.addChild(eyeL);
headPivot.addChild(eyeR);
headPivot.addChild(earL);
headPivot.addChild(earR);

const tailPivot = doc.createNode('TailPivot').setTranslation([0, 0.78, -0.78]);
rootNode.addChild(tailPivot);
const tailNode = doc.createNode('Tail').setMesh(tailMesh);
setNodeRotationFromEuler(tailNode, -0.35, 0, 0);
tailPivot.addChild(tailNode);

const legPositions = {
  Leg_FL_Pivot: [0.32, 0.48, 0.42],
  Leg_FR_Pivot: [-0.32, 0.48, 0.42],
  Leg_BL_Pivot: [0.32, 0.48, -0.42],
  Leg_BR_Pivot: [-0.32, 0.48, -0.42],
};

const legPivotMap = {};

Object.entries(legPositions).forEach(([name, translation]) => {
  const pivot = doc.createNode(name).setTranslation(translation);
  legPivotMap[name] = pivot;
  const legNode = doc
    .createNode(name.replace('_Pivot', ''))
    .setMesh(legMesh)
    .setTranslation([0, -0.18, 0]);
  pivot.addChild(legNode);
  rootNode.addChild(pivot);
});

// --- アニメーション作成 ---
function addQuaternionTrack(animation, node, keyframes) {
  const times = new Float32Array(keyframes.map((k) => k.time));
  const rotations = new Float32Array(keyframes.length * 4);
  keyframes.forEach((keyframe, index) => {
    const quat = keyframe.quaternion
      ? keyframe.quaternion
      : new THREE.Quaternion().setFromEuler(
          new THREE.Euler(keyframe.rotation[0], keyframe.rotation[1], keyframe.rotation[2])
        );
    rotations.set([quat.x, quat.y, quat.z, quat.w], index * 4);
  });

  const sampler = doc
    .createAnimationSampler(`${animation.getName()}_${node.getName()}_sampler`)
    .setInput(accessorFromArray(`${animation.getName()}_${node.getName()}_time`, times, Accessor.Type.SCALAR))
    .setOutput(accessorFromArray(`${animation.getName()}_${node.getName()}_rot`, rotations, Accessor.Type.VEC4))
    .setInterpolation('LINEAR');

  const channel = doc.createAnimationChannel(`${animation.getName()}_${node.getName()}_channel`);
  channel.setTargetNode(node).setTargetPath('rotation').setSampler(sampler);
  animation.addSampler(sampler).addChannel(channel);
}

function createAnimation(name, tracks) {
  const animation = doc.createAnimation(name);
  tracks.forEach(({ node, keyframes }) => addQuaternionTrack(animation, node, keyframes));
}

createAnimation('idle', [
  {
    node: headPivot,
    keyframes: [
      { time: 0, rotation: [0.05, 0, 0] },
      { time: 1.2, rotation: [-0.03, 0.03, 0] },
      { time: 2.4, rotation: [0.05, 0, 0] },
    ],
  },
  {
    node: tailPivot,
    keyframes: [
      { time: 0, rotation: [0, 0.15, 0] },
      { time: 1.2, rotation: [0, -0.15, 0] },
      { time: 2.4, rotation: [0, 0.15, 0] },
    ],
  },
]);

createAnimation('walk', [
  {
    node: legPivotMap.Leg_FL_Pivot,
    keyframes: [
      { time: 0, rotation: [0.45, 0, 0] },
      { time: 0.4, rotation: [-0.45, 0, 0] },
      { time: 0.8, rotation: [0.45, 0, 0] },
    ],
  },
  {
    node: legPivotMap.Leg_FR_Pivot,
    keyframes: [
      { time: 0, rotation: [-0.45, 0, 0] },
      { time: 0.4, rotation: [0.45, 0, 0] },
      { time: 0.8, rotation: [-0.45, 0, 0] },
    ],
  },
  {
    node: legPivotMap.Leg_BL_Pivot,
    keyframes: [
      { time: 0, rotation: [-0.4, 0, 0] },
      { time: 0.4, rotation: [0.4, 0, 0] },
      { time: 0.8, rotation: [-0.4, 0, 0] },
    ],
  },
  {
    node: legPivotMap.Leg_BR_Pivot,
    keyframes: [
      { time: 0, rotation: [0.4, 0, 0] },
      { time: 0.4, rotation: [-0.4, 0, 0] },
      { time: 0.8, rotation: [0.4, 0, 0] },
    ],
  },
  {
    node: headPivot,
    keyframes: [
      { time: 0, rotation: [0.1, 0.05, 0] },
      { time: 0.4, rotation: [-0.1, -0.05, 0] },
      { time: 0.8, rotation: [0.1, 0.05, 0] },
    ],
  },
  {
    node: tailPivot,
    keyframes: [
      { time: 0, rotation: [0, 0.25, 0] },
      { time: 0.4, rotation: [0, -0.25, 0] },
      { time: 0.8, rotation: [0, 0.25, 0] },
    ],
  },
]);

createAnimation('tail_wag', [
  {
    node: tailPivot,
    keyframes: [
      { time: 0, rotation: [0, 0.5, 0] },
      { time: 0.25, rotation: [0, -0.5, 0] },
      { time: 0.5, rotation: [0, 0.5, 0] },
    ],
  },
]);

createAnimation('sit', [
  {
    node: legPivotMap.Leg_BL_Pivot,
    keyframes: [
      { time: 0, rotation: [0, 0, 0] },
      { time: 1.0, rotation: [-1.1, 0, 0] },
      { time: 1.6, rotation: [-1.1, 0, 0] },
    ],
  },
  {
    node: legPivotMap.Leg_BR_Pivot,
    keyframes: [
      { time: 0, rotation: [0, 0, 0] },
      { time: 1.0, rotation: [-1.1, 0, 0] },
      { time: 1.6, rotation: [-1.1, 0, 0] },
    ],
  },
  {
    node: legPivotMap.Leg_FL_Pivot,
    keyframes: [
      { time: 0, rotation: [0, 0, 0] },
      { time: 1.0, rotation: [0.25, 0, 0] },
      { time: 1.6, rotation: [0.25, 0, 0] },
    ],
  },
  {
    node: legPivotMap.Leg_FR_Pivot,
    keyframes: [
      { time: 0, rotation: [0, 0, 0] },
      { time: 1.0, rotation: [0.25, 0, 0] },
      { time: 1.6, rotation: [0.25, 0, 0] },
    ],
  },
  {
    node: bodyNode,
    keyframes: [
      { time: 0, rotation: [0, 0, 0] },
      { time: 1.0, rotation: [-0.4, 0, 0] },
      { time: 1.6, rotation: [-0.4, 0, 0] },
    ],
  },
  {
    node: headPivot,
    keyframes: [
      { time: 0, rotation: [0, 0, 0] },
      { time: 1.0, rotation: [0.35, 0.1, 0] },
      { time: 1.6, rotation: [0.35, 0.1, 0] },
    ],
  },
]);

async function main() {
  await fs.mkdir(path.dirname(OUTPUT_PATH), { recursive: true });
  const io = new NodeIO();
  const glbData = await io.writeBinary(doc);
  await fs.writeFile(OUTPUT_PATH, glbData);
  console.log(`GLBを生成しました: ${OUTPUT_PATH}`);
}

main().catch((error) => {
  console.error('犬GLB生成中にエラーが発生しました', error);
  process.exitCode = 1;
});

File: dog-3d-app/tools/generate_dog_glb.py
"""
このリポジトリのCI環境では未実行。Blenderローカルで `blender -b -P tools/generate_dog_glb.py` を実行してください。
"""

import bpy
from math import pi
from mathutils import Euler, Quaternion, Vector

# 既存データをクリーンアップ
bpy.ops.object.select_all(action='SELECT')
bpy.ops.object.delete(use_global=False)

scene = bpy.context.scene
scene.frame_start = 0
scene.frame_end = 160

# --- マテリアル作成 ---
def make_material(name, hex_color):
    material = bpy.data.materials.new(name)
    material.use_nodes = True
    nodes = material.node_tree.nodes
    principled = nodes.get('Principled BSDF')
    rgb = [(hex_color >> 16 & 0xFF) / 255.0, (hex_color >> 8 & 0xFF) / 255.0, (hex_color & 0xFF) / 255.0, 1.0]
    principled.inputs['Base Color'].default_value = rgb
    principled.inputs['Metallic'].default_value = 0.0
    principled.inputs['Roughness'].default_value = 0.6
    return material

materials = {
    'Dog_Body_Mat': make_material('Dog_Body_Mat', 0x8A6547),
    'Dog_Ear_Mat': make_material('Dog_Ear_Mat', 0x5A3D27),
    'Dog_Nose_Mat': make_material('Dog_Nose_Mat', 0x1A1A1A),
    'Dog_Eye_Mat': make_material('Dog_Eye_Mat', 0x101010),
    'Dog_Paw_Mat': make_material('Dog_Paw_Mat', 0xD8B99B),
}

# --- ノード階層の構築 ---
root = bpy.data.objects.new('DogRoot', None)
scene.collection.objects.link(root)
root.empty_display_type = 'PLAIN_AXES'

body_mesh = bpy.data.meshes.new('BodyMesh')
bpy.ops.mesh.primitive_uv_sphere_add(segments=12, ring_count=10, radius=0.4, location=(0, 0.58, 0))
body_obj = bpy.context.active_object
body_obj.name = 'Body'
body_obj.data = bpy.data.meshes.new_from_object(body_obj, preserve_all_data_layers=True, depsgraph=bpy.context.evaluated_depsgraph_get())
bpy.ops.object.delete()
body_obj.data.materials.append(materials['Dog_Body_Mat'])
scene.collection.objects.link(body_obj)
body_obj.parent = root

bpy.ops.mesh.primitive_uv_sphere_add(segments=16, ring_count=12, radius=0.35, location=(0, 1.03, 0.92))
head_obj = bpy.context.active_object
head_obj.name = 'HeadMeshTmp'

bpy.ops.mesh.primitive_cube_add(size=0.34, location=(0, 1.0, 1.3))
snout_box = bpy.context.active_object
snout_box.name = 'SnoutBoxTmp'

bpy.ops.mesh.primitive_uv_sphere_add(segments=12, ring_count=8, radius=0.1, location=(0, 1.03, 1.55))
nose_obj = bpy.context.active_object
nose_obj.name = 'NoseTmp'

bpy.ops.object.select_all(action='DESELECT')
head_obj.select_set(True)
nose_obj.select_set(True)
snout_box.select_set(True)
bpy.context.view_layer.objects.active = head_obj
bpy.ops.object.join()
head_obj = bpy.context.active_object
head_obj.name = 'HeadCombined'
head_obj.data.materials.append(materials['Dog_Body_Mat'])
head_obj.data.materials.append(materials['Dog_Nose_Mat'])
for poly in head_obj.data.polygons:
    if head_obj.data.vertices[poly.vertices[0]].co.z > 1.35:
        poly.material_index = 1
scene.collection.objects.link(head_obj)

head_pivot = bpy.data.objects.new('HeadPivot', None)
head_pivot.empty_display_type = 'PLAIN_AXES'
head_pivot.location = (0, 0.95, 0.68)
scene.collection.objects.link(head_pivot)
head_pivot.parent = root

head = bpy.data.objects.new('Head', head_obj.data.copy())
head.location = (0, 0.08, 0.24)
head.rotation_euler = (-0.05, 0.0, 0.0)
head.parent = head_pivot
scene.collection.objects.link(head)

bpy.ops.mesh.primitive_uv_sphere_add(segments=8, ring_count=6, radius=0.06, location=(0, 0.0, 0.0))
_eye_mesh = bpy.context.active_object.data.copy()
bpy.ops.object.delete()

def create_eye(name, offset):
    eye = bpy.data.objects.new(name, _eye_mesh.copy())
    eye.location = offset
    eye.data.materials.append(materials['Dog_Eye_Mat'])
    eye.parent = head_pivot
    scene.collection.objects.link(eye)

create_eye('Eye_L', Vector((-0.16, 0.05, 0.93)))
create_eye('Eye_R', Vector((0.16, 0.05, 0.93)))

bpy.ops.mesh.primitive_cone_add(vertices=4, radius1=0.14, depth=0.5, location=(0, 0, 0))
ear_mesh = bpy.context.active_object.data.copy()
bpy.ops.object.delete()

def create_ear(name, offset, rotation):
    ear = bpy.data.objects.new(name, ear_mesh.copy())
    ear.location = offset
    ear.rotation_euler = rotation
    ear.data.materials.append(materials['Dog_Ear_Mat'])
    ear.parent = head_pivot
    scene.collection.objects.link(ear)

create_ear('Ear_L', Vector((-0.22, 0.15, 0.56)), Euler((-0.4, 0.0, 0.35)))
create_ear('Ear_R', Vector((0.22, 0.15, 0.56)), Euler((-0.4, 0.0, -0.35)))

bpy.ops.mesh.primitive_cylinder_add(vertices=6, radius=0.11, depth=0.7, location=(0, 0.78, -0.78))
tail_base = bpy.context.active_object
bpy.ops.transform.rotate(value=pi / 2, orient_axis='X')
bpy.ops.transform.translate(value=(0, 0, -0.35))
tail_mesh = tail_base.data.copy()
bpy.ops.object.delete()

tail_pivot = bpy.data.objects.new('TailPivot', None)
tail_pivot.empty_display_type = 'PLAIN_AXES'
tail_pivot.location = (0, 0.78, -0.78)
tail_pivot.parent = root
scene.collection.objects.link(tail_pivot)

tail = bpy.data.objects.new('Tail', tail_mesh.copy())
tail.rotation_euler = Euler((-0.35, 0.0, 0.0))
tail.parent = tail_pivot
scene.collection.objects.link(tail)
tail.data.materials.append(materials['Dog_Body_Mat'])

bpy.ops.mesh.primitive_uv_sphere_add(segments=10, ring_count=6, radius=0.1)
leg_core_mesh = bpy.context.active_object.data.copy()
bpy.ops.object.delete()

bpy.ops.mesh.primitive_cylinder_add(vertices=6, radius=0.13, depth=0.12)
paw_mesh = bpy.context.active_object.data.copy()
bpy.ops.object.delete()

def create_leg(name, pivot_pos):
    pivot = bpy.data.objects.new(f"{name}_Pivot", None)
    pivot.location = pivot_pos
    pivot.empty_display_type = 'PLAIN_AXES'
    pivot.parent = root
    scene.collection.objects.link(pivot)

    leg_obj = bpy.data.objects.new(name, leg_core_mesh.copy())
    leg_obj.location = (0, -0.18, 0)
    leg_obj.data.materials.append(materials['Dog_Body_Mat'])
    scene.collection.objects.link(leg_obj)
    paw_obj = bpy.data.objects.new(f"{name}_PawTemp", paw_mesh.copy())
    paw_obj.location = (0, -0.3, 0)
    paw_obj.data.materials.append(materials['Dog_Paw_Mat'])
    scene.collection.objects.link(paw_obj)

    bpy.ops.object.select_all(action='DESELECT')
    leg_obj.select_set(True)
    paw_obj.select_set(True)
    bpy.context.view_layer.objects.active = leg_obj
    bpy.ops.object.join()

    leg_obj.parent = pivot

for name, pos in {
    'Leg_FL': Vector((0.32, 0.48, 0.42)),
    'Leg_FR': Vector((-0.32, 0.48, 0.42)),
    'Leg_BL': Vector((0.32, 0.48, -0.42)),
    'Leg_BR': Vector((-0.32, 0.48, -0.42)),
}.items():
    create_leg(name, pos)

# --- アニメーションのセットアップ ---

def ensure_animation_data(obj):
    if obj.animation_data is None:
        obj.animation_data_create()
    return obj.animation_data


def create_action(name):
    action = bpy.data.actions.new(name)
    action.use_fake_user = True
    return action


def record_rotation(action, obj_name, keyframes):
    obj = bpy.data.objects[obj_name]
    obj.rotation_mode = 'XYZ'
    anim_data = ensure_animation_data(obj)
    anim_data.action = action
    for frame, euler in keyframes:
        obj.rotation_euler = Euler(euler)
        obj.keyframe_insert(data_path='rotation_euler', frame=frame)
    anim_data.action = None

idle_action = create_action('idle')
record_rotation(idle_action, 'HeadPivot', [(0, (0.05, 0.0, 0.0)), (30, (-0.03, 0.03, 0.0)), (60, (0.05, 0.0, 0.0))])
record_rotation(idle_action, 'TailPivot', [(0, (0.0, 0.15, 0.0)), (30, (0.0, -0.15, 0.0)), (60, (0.0, 0.15, 0.0))])

walk_action = create_action('walk')
record_rotation(walk_action, 'Leg_FL_Pivot', [(0, (0.45, 0.0, 0.0)), (12, (-0.45, 0.0, 0.0)), (24, (0.45, 0.0, 0.0))])
record_rotation(walk_action, 'Leg_FR_Pivot', [(0, (-0.45, 0.0, 0.0)), (12, (0.45, 0.0, 0.0)), (24, (-0.45, 0.0, 0.0))])
record_rotation(walk_action, 'Leg_BL_Pivot', [(0, (-0.4, 0.0, 0.0)), (12, (0.4, 0.0, 0.0)), (24, (-0.4, 0.0, 0.0))])
record_rotation(walk_action, 'Leg_BR_Pivot', [(0, (0.4, 0.0, 0.0)), (12, (-0.4, 0.0, 0.0)), (24, (0.4, 0.0, 0.0))])
record_rotation(walk_action, 'HeadPivot', [(0, (0.1, 0.05, 0.0)), (12, (-0.1, -0.05, 0.0)), (24, (0.1, 0.05, 0.0))])
record_rotation(walk_action, 'TailPivot', [(0, (0.0, 0.25, 0.0)), (12, (0.0, -0.25, 0.0)), (24, (0.0, 0.25, 0.0))])

tail_action = create_action('tail_wag')
record_rotation(tail_action, 'TailPivot', [(0, (0.0, 0.5, 0.0)), (10, (0.0, -0.5, 0.0)), (20, (0.0, 0.5, 0.0))])

sit_action = create_action('sit')
record_rotation(sit_action, 'Leg_BL_Pivot', [(0, (0.0, 0.0, 0.0)), (40, (-1.1, 0.0, 0.0)), (60, (-1.1, 0.0, 0.0))])
record_rotation(sit_action, 'Leg_BR_Pivot', [(0, (0.0, 0.0, 0.0)), (40, (-1.1, 0.0, 0.0)), (60, (-1.1, 0.0, 0.0))])
record_rotation(sit_action, 'Leg_FL_Pivot', [(0, (0.0, 0.0, 0.0)), (40, (0.25, 0.0, 0.0)), (60, (0.25, 0.0, 0.0))])
record_rotation(sit_action, 'Leg_FR_Pivot', [(0, (0.0, 0.0, 0.0)), (40, (0.25, 0.0, 0.0)), (60, (0.25, 0.0, 0.0))])
record_rotation(sit_action, 'Body', [(0, (0.0, 0.0, 0.0)), (40, (-0.4, 0.0, 0.0)), (60, (-0.4, 0.0, 0.0))])
record_rotation(sit_action, 'HeadPivot', [(0, (0.0, 0.0, 0.0)), (40, (0.35, 0.1, 0.0)), (60, (0.35, 0.1, 0.0))])

# --- GLB書き出し設定 ---
export_path = bpy.path.abspath('//public/models/dog.glb')
bpy.ops.export_scene.gltf(
    filepath=export_path,
    export_format='GLB',
    export_apply=True,
    export_nla_strips=True,
    export_anim_single_armature=False,
    export_frame_range=False,
    export_selected=False,
    export_current_frame=False,
)
print(f"GLBを書き出しました: {export_path}")
